File: libs/java/dal/db/dal-db/src/main/java/ca/bc/gov/tno/dal/db/entities/ContentCategory.java
package ca.bc.gov.tno.dal.db.entities;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.Id;
import javax.persistence.IdClass;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.Table;

import ca.bc.gov.tno.dal.db.AuditColumns;

/**
 * ContentCategory class, provides a way to manage content categories.
 */
@Entity
@IdClass(ContentCategoryPK.class)
@Table(name = "content_category", schema = "public")
public class ContentCategory extends AuditColumns {
  /**
   * Primary key to identify the content category.
   * Foreign key to content.
   */
  @Id
  @Column(name = "content_id", nullable = false)
  private int contentId;

  /**
   * The content reference.
   */
  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "content_id", insertable = false, updatable = false)
  private Content content;

  /**
   * Primary key to identify the content category.
   * Foreign key to category .
   */
  @Id
  @Column(name = "category_id", nullable = false)
  private int categoryId;

  /**
   * The category reference.
   */
  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "category_id", insertable = false, updatable = false)
  private Category category;

  /**
   * Score of category.
   */
  @Column(name = "score", nullable = false)
  private int score;

  /**
   * Creates a new instance of a ContentCategory object.
   */
  public ContentCategory() {

  }

  /**
   * Creates a new instance of a ContentCategory object, initializes with
   * specified
   * parameters.
   * 
   * @param content  Content object
   * @param category Category object
   * @param score    Category score
   */
  public ContentCategory(Content content, Category category, int score) {
    if (content == null)
      throw new NullPointerException("Parameter 'content' cannot be null.");
    if (category == null)
      throw new NullPointerException("Parameter 'category' cannot be null.");

    this.content = content;
    this.contentId = content.getId();
    this.category = category;
    this.categoryId = category.getId();
    this.score = score;
  }

  /**
   * @return int return the contentId
   */
  public int getContentId() {
    return contentId;
  }

  /**
   * @param contentId the contentId to set
   */
  public void setContentId(int contentId) {
    this.contentId = contentId;
  }

  /**
   * @return Content return the content
   */
  public Content getContent() {
    return content;
  }

  /**
   * @param content the content to set
   */
  public void setContent(Content content) {
    this.content = content;
  }

  /**
   * @return int return the categoryId
   */
  public int getCategoryId() {
    return categoryId;
  }

  /**
   * @param categoryId the categoryId to set
   */
  public void setCategoryId(int categoryId) {
    this.categoryId = categoryId;
  }

  /**
   * @return Category return the category
   */
  public Category getCategory() {
    return category;
  }

  /**
   * @param category the category to set
   */
  public void setCategory(Category category) {
    this.category = category;
  }

  /**
   * @return int return the score
   */
  public int getScore() {
    return score;
  }

  /**
   * @param score the score to set
   */
  public void setScore(int score) {
    this.score = score;
  }

}

File: libs/net/entities/ContentCategory.cs
using System.ComponentModel.DataAnnotations.Schema;

namespace TNO.Entities;

/// <summary>
/// ContentCategory class, provides an entity model to link (many-to-many) content with categories.
/// </summary>
[Table("content_category")]
public class ContentCategory : AuditColumns, IEquatable<ContentCategory>
{
    #region Properties
    /// <summary>
    /// get/set - Primary key and foreign key to content.
    /// </summary>
    [Column("content_id")]
    public long ContentId { get; set; }

    /// <summary>
    /// get/set - The content.
    /// </summary>
    public virtual Content? Content { get; set; }

    /// <summary>
    /// get/set - Primary key and foreign key to category.
    /// </summary>
    [Column("category_id")]
    public int CategoryId { get; set; }

    /// <summary>
    /// get/set - The category.
    /// </summary>
    public virtual Category? Category { get; set; }

    /// <summary>
    /// get/set - The score of the category.
    /// </summary>
    [Column("score")]
    public int Score { get; set; }
    #endregion

    #region Constructors
    protected ContentCategory() { }

    public ContentCategory(long contentId, int categoryId, int score)
    {
        this.ContentId = contentId;
        this.CategoryId = categoryId;
        this.Score = score;
    }

    public ContentCategory(Content content, Category category, int score)
    {
        this.ContentId = content?.Id ?? throw new ArgumentNullException(nameof(content));
        this.Content = content;
        this.CategoryId = category?.Id ?? throw new ArgumentNullException(nameof(category));
        this.Category = category;
        this.Score = score;
    }
    #endregion

    #region Methods
    public bool Equals(ContentCategory? other)
    {
        if (other == null) return false;
        return this.ContentId == other.ContentId && this.CategoryId == other.CategoryId;
    }

    public override bool Equals(object? obj) => Equals(obj as ContentCategory);
    public override int GetHashCode() => (this.ContentId, this.CategoryId).GetHashCode();
    #endregion
}
